js/src/rive.ts

```diff
@@ -0,0 +1,965 @@
+  }
+}
+
+// #endregion
+
+// #region animations
+
+// Wraps animations and instances from the runtime and keeps track of playback
+// state
+export class Animation {  
+  public loopCount: number = 0;
+  public paused: boolean = false;
+
+  constructor(private animation: any, public readonly instance: any) {}
+
+  // Returns the animation's name
+  public get name(): string {
+    return this.animation.name;
+  }
+
+  // Returns the animation's loop type
+  public get loopValue(): number {
+    return this.animation.loopValue;
+  } 
+}
+
+// #endregion
+
+// #region events
+
+// Events that Rive fires
+export enum EventType {
+  Load      = 'load',
+  LoadError = 'loaderror',
+  Play      = 'play',
+  Pause     = 'pause',
+  Stop      = 'stop',
+  Loop      = 'loop',
+}
+
+// Event fired by Rive
+export interface Event {
+  type: EventType,
+  data?: string | string[],
+}
+
+// Callback type for event listeners
+export type EventCallback = (event: Event) => void;
+
+// An event listener
+export interface EventListener {
+  type: EventType,
+  callback: EventCallback, 
+}
+
+// Manages Rive events and listeners
+class EventManager {
+
+  constructor(private listeners: EventListener[] = []) {}
+
+  // Gets listeners of specified type
+  private getListeners(type: EventType): EventListener[] {
+    return this.listeners.filter(e => e.type === type);
+  }
+
+  // Adds a listener
+  public add(listener: EventListener): void {
+    if (!this.listeners.includes(listener)) {
+      this.listeners.push(listener);
+    }
+  }
+
+  // Removes listener
+  public remove(listener: EventListener): void {
+    const index = this.listeners.indexOf(listener, 0);
+    if (index > -1) {
+      this.listeners.splice(index, 1);
+    }
+  }
+
+  // Fires an event
+  public fire(event: Event, ignoreDuplicate = false): void {
+    const eventListeners = this.getListeners(event.type);
+    eventListeners.forEach(
+      listener => listener.callback(event)
+    );
+  }
+}
+
+// #endregion
+
+// #region Manages a queue of tasks
+
+// A task in the queue; will fire the action when the queue is processed; will
+// also optionally fire an event.
+export interface Task {
+  action: ActionCallback,
+  event?: Event,
+}
+
+// Callback type for task actions
+export type ActionCallback = () => void;
+
+// Manages a queue of tasks
+class TaskQueueManager {
+  private queue: Task[] = [];
+
+  constructor(private eventManager: EventManager) {}
+
+  // Adds a task top the queue
+  public add(task: Task): void {
+    this.queue.push(task);
+  }
+
+  // Processes all tasks in the queue
+  public process(): void {
+    while (this.queue.length > 0) {
+      const task = this.queue.shift();
+      task.action();
+      if (task.event) {
+        this.eventManager.fire(task.event);
+      }
+    }
+  }
+}
+
+// #region Rive
+
+// Interface for the Rive static method contructor
+export interface RiveParameters {
+  canvas: HTMLCanvasElement, // canvas is required
```

Comment:
Technically it can also be an `OffscreenCanvas` that runs in a webworker no ?

Reply:
Not sure; I'll check.